tests/test_proxy.py

```diff
@@ -1,29 +1,289 @@
+        (
+            "host/proxy/path",
+            "http://127.0.0.1:99",
+            {"test": "test2"},
+            {
+                "routespec": "host/proxy/path",
+                "target": "http://127.0.0.1:99",
+                "data": json.dumps({"test": "test2"}),
+            },
+        ),  # Host-based routing
+    ],
+)
+async def test_get_route(proxy, routespec, target, data, expected_output):
+    add_route_with_etcdctl(routespec, target, data)
+    route = await proxy.get_route(routespec)
+    cleanup_test_route(routespec, target, data)
+
+    assert route == expected_output
+    await proxy.stop()
 
 
 async def test_get_all_routes(proxy):
-    with pytest.raises(NotImplementedError):
-        routes = await proxy.get_all_routes()
-    # TODO: test the routes
+    routespec = ["/proxy/path1", "/proxy/path2", "host/proxy/path"]
+    target = ["http://127.0.0.1:990", "http://127.0.0.1:909", "http://127.0.0.1:999"]
+    data = [{"test": "test1"}, {}, {"test": "test2"}]
+    dict_keys = ["routespec", "target", "data"]
 
+    expected_output = {
+        routespec[0]: {
+            "routespec": routespec[0],
+            "target": target[0],
+            "data": json.dumps(data[0]),
+        },
+        routespec[1]: {
+            "routespec": routespec[1],
+            "target": target[1],
+            "data": json.dumps(data[1]),
+        },
+        routespec[2]: {
+            "routespec": routespec[2],
+            "target": target[2],
+            "data": json.dumps(data[2]),
+        },
+    }
 
-async def test_delete_route(proxy):
-    with pytest.raises(NotImplementedError):
-        await proxy.delete_route("/prefix")
+    add_route_with_etcdctl(routespec[0], target[0], data[0])
+    add_route_with_etcdctl(routespec[1], target[1], data[1])
+    add_route_with_etcdctl(routespec[2], target[2], data[2])
+    routes = await proxy.get_all_routes()
+    try:
+        assert routes == expected_output
+    finally:
+        cleanup_test_route(routespec[0], target[0], data[0])
+        cleanup_test_route(routespec[1], target[1], data[1])
+        cleanup_test_route(routespec[2], target[2], data[2])
+        await proxy.stop()
 
 
-async def test_get_route(proxy):
-    with pytest.raises(NotImplementedError):
-        route = await proxy.get_route("/prefix")
-    # TODO: test the route
+async def test_etcd_routing(proxy):
+    routespec = ["/", "/user/first", "/user/second"]
+    target = ["http://127.0.0.1:9000", "http://127.0.0.1:9090", "http://127.0.0.1:9099"]
+    data = [{}, {}, {}]
+    try:
+        await proxy.add_route(routespec[0], target[0], data[0])
+        await proxy.add_route(routespec[1], target[1], data[1])
+        await proxy.add_route(routespec[2], target[2], data[2])
+        """ Force the proxy reload its configuration from etcd, otherwise
+            we have no other way to ensure traefik loaded the newly added routes
+            from etcd.
+        """
+        proxy._stop_traefik()
```

Comment:
In a later PR, we can put this starting and stopping traefik into a fixture to make it more reusable